queen-of-versions/src/main/java/co/infinum/queenofversions/GoogleInAppUpdateCallback.java

```diff
@@ -72,44 +165,135 @@ public void onStateUpdate(InstallState installState) {
-            flexibleStateListener.onFailed(new GoogleInAppUpdateException(InAppUpdateError.INTERNAL_ERROR));
+            flexibleStateListener.onFailed(new GoogleInAppUpdateException(INTERNAL_ERROR));
         } else if (installState.installErrorCode() == InstallErrorCode.ERROR_INVALID_REQUEST) {
-            flexibleStateListener.onFailed(new GoogleInAppUpdateException(InAppUpdateError.INVALID_REQUEST));
+            flexibleStateListener.onFailed(new GoogleInAppUpdateException(INVALID_REQUEST));
         } else if (installState.installErrorCode() == InstallErrorCode.ERROR_UNKNOWN) {
-            flexibleStateListener.onFailed(new GoogleInAppUpdateException(InAppUpdateError.ERROR_UNKNOWN));
+            flexibleStateListener.onFailed(new GoogleInAppUpdateException(ERROR_UNKNOWN));
+        }
+    }
+
+    /**
+     * Method that is called every time we check for an update. Whether there is or isn't an update on {@link PrinceOfVersions} we are
+     * still going to check with Google to be sure.
+     *
+     * @param isMandatory       Determines if the update we are checking for is MANDATORY or OPTIONAL
+     * @param princeVersionCode Version code that we got from {@link PrinceOfVersions} after parsing JSON file
+     */
+    private void checkWithGoogleForAnUpdate(boolean isMandatory, String princeVersionCode) {
+        googleAppUpdater.initGoogleUpdate(isMandatory, princeVersionCode);
+    }
+
+    /**
+     * Method that is used to check what kind of update do we want to invoke on Google Play Store. Depending on version codes
+     * on JSON file, application  and Google Play Store, we can invoke FLEXIBLE or IMMEDIATE update on Google Play Store.
+     *
+     * @param princeVersionCode Version code that we got from {@link PrinceOfVersions} after parsing JSON file
+     * @param googleVersionCode Version code that we got from Google Play Store
+     * @param isMandatory       Determines whether the update we have is MANDATORY or OPTIONAL
+     * @return Returns an {@link AppUpdateType} depending on version codes we have
+     */
+    private int checkVersionCode(String princeVersionCode, int googleVersionCode, boolean isMandatory) {
+        int princeOfVersionsCode;
+        if (princeVersionCode == null) {
+            return AppUpdateType.FLEXIBLE;
+        } else {
+            princeOfVersionsCode = Integer.parseInt(princeVersionCode);
+        }
+
+        if (princeOfVersionsCode <= googleVersionCode && isMandatory) {
+            if (appVersionCode > princeOfVersionsCode) {
+                return AppUpdateType.FLEXIBLE;
+            } else {
+                return AppUpdateType.IMMEDIATE;
+            }
+        } else if (princeOfVersionsCode > googleVersionCode) {
+            if (appVersionCode
+                > googleVersionCode) { //This if shouldn't be possible in production. App version can't be lower than Google Play version.
+                return UNSUPPORTED_VERSION;
```

Comment:
Let's make an enum for update type. It will be much cleaner than comparing int with some code 3. Also, what if google in future change value of FLEXIBLE or IMMEDIATE to 3. It's better to not depend on that.